pyadjoint/optimization/tao_solver.py

```diff
@@ -115,34 +113,41 @@ def from_petsc(self, y, X):
 
         if y_a.shape != (self.n,):
             raise ValueError("Invalid shape")
-        if len(X) != len(self.indices):
+        if len(X) != len(self._index_concatenated_vecs):
             raise ValueError("Invalid length")
-
-        for (i0, i1), x in zip(self.indices, X):
-            _, x_i1 = x._ad_assign_numpy(x, y_a, offset=i0)
-            if i1 != x_i1:
-                raise ValueError("Invalid index")
+        for i, x in enumerate(X):
+            x._ad_petsc_vec().setArray(y_a[self._index_concatenated_vecs[i]])
 
     def to_petsc(self, x, Y):
         """Copy data from variables to a :class:`petsc4py.PETSc.Vec`.
 
         Args:
             x (petsc4py.PETSc.Vec): The output :class:`petsc4py.PETSc.Vec`.
-            Y (numbers.Complex, OverloadedType or Sequence[OverloadedType]):
+            Y (OverloadedType or Sequence[OverloadedType]):
                 Values for input variables.
         """
 
         Y = Enlist(Y)
         if len(Y) != len(self.indices):
             raise ValueError("Invalid length")
+        x.setArray(
+            self.concatenate_vecs([y._ad_petsc_vec() for y in Y]).getArray()
+        )
 
-        x_a = np.zeros(self.n, dtype=PETSc.ScalarType)
-        for (i0, i1), y in zip(self.indices, Y):
-            if isinstance(y, numbers.Complex):
-                x_a[i0:i1] = y
-            else:
-                x_a[i0:i1] = y._ad_to_list(y)
-        x.setArray(x_a)
+    def concatenate_vecs(self, vecs):
+        """Concatenate Vecs into a single Vec.
+
+        Args:
+            Y (`petsc4py.PETSc.Vec` or sequence of `petsc4py.PETSc.Vec`):
+            The input Vecs to concatenate.
+        """
+        if not all(isinstance(vec, PETSc.Vec) for vec in vecs):
+            raise ValueError("vecs should be a sequence of PETSc.Vec objects")
+
+        vecs = Enlist(vecs)
+        concatenated_vec, set_indexes = PETSc.Vec().concatenate(vecs)
+        self._index_concatenated_vecs = [set_indexes[i].array for i in range(len(vecs))]
```

Comment:
If the `IS` objects are retained then `VecGetSubVector`/`VecRestoreSubVector` can be used.